### PR TITLE
Fix unit "Not Sure" option

### LINKED ISSUE #132: Unit "Not Sure" option is broken
If you select the "Not Sure" unit option on the filter officers page and click submit, the search page will not show any unit options you checked.

This is because we filter the formdata here:
https://github.com/OrcaCollective/OpenOversight/blob/e041f7f0b6e10d50e52b1ce13fa39037c01b0ae3/OpenOversight/app/main/views.py#L676-L697

Before we add the "Not Sure" option to the choices dropdown here:
https://github.com/OrcaCollective/OpenOversight/blob/e041f7f0b6e10d50e52b1ce13fa39037c01b0ae3/OpenOversight/app/main/views.py#L714-L719

If we add "Not Sure" to `unit_choices`, that should be passed to `filter_by_form` as expected here:
https://github.com/OrcaCollective/OpenOversight/blob/e041f7f0b6e10d50e52b1ce13fa39037c01b0ae3/OpenOversight/app/utils.py#L354-L365

### PR DESCRIPTION
**Pull request by sea-kelp**
## Description of Changes
Fixes #132.

Fixes a bug where selecting "Not Sure" in the Unit dropdown would cause the officer query to ignore any unit selections.

## Notes for Deployment
None!

## Screenshots (if appropriate)
Selecting "Not Sure" and "Community Outreach"

### Before
The search ignores the selected units

### After
The search only lists officers with the "Community Outreach" unit or no unit

## Tests and linting

 - [x] I have rebased my changes on `main`

 - [x] `just lint` passes

 - [x] `just test` passes

### CODE REVIEW

File: OpenOversight/app/main/views.py [hunk 1]
```diff
@@ -673,7 +673,7 @@ def list_officer(
     ):
         form_data["gender"] = genders
 
-    unit_choices = [
+    unit_choices = ["Not Sure"] + [
```
Comment (AetherUnbound) [inline OpenOversight/app/main/views.py:676]: Why was this change necessary? It looks like it'd produce the same results 🤔

File: OpenOversight/app/utils.py [hunk 2]
```diff
@@ -361,17 +362,24 @@ def filter_by_form(form_data, officer_query, department_id=None):
         ]
 
         if "Not Sure" in form_data["unit"]:
-            # Convert "Not Sure" to None, so the resulting SQL query allows NULL values
-            form_data["unit"].append(None)
+            include_null_unit = True
```
Comment (AetherUnbound) [inline OpenOversight/app/utils.py:365]: Ha, looks like I introduced this in #91 😅

File: OpenOversight/tests/conftest.py [hunk 3]
```diff
@@ -131,12 +131,17 @@ def generate_officer():
     )
 
 
-def build_assignment(officer: Officer, units: List[Unit], jobs: Job):
+def build_assignment(officer: Officer, units: List[Optional[Unit]], jobs: Job):
+    unit = random.choice(units)
+    if unit:
+        unit_id = unit.id
+    else:
+        unit_id = None
```
Comment (AetherUnbound) [inline OpenOversight/tests/conftest.py:139]: Small simplification
```suggestion
    unit_id = unit.id if unit else None
```

File: OpenOversight/tests/routes/test_officer_and_department.py [hunk 4]
```diff
@@ -1830,8 +1837,14 @@ def test_ac_can_upload_photos_of_dept_officers(
...
-        crop_mock = MagicMock(return_value=Image.query.first())
-        upload_mock = MagicMock(return_value=Image.query.first())
+        # Filter out images that the officer is already tagged in
+        officer_faces = Face.query.filter_by(officer_id=officer.id).all()
+        image = Image.query.filter(
+            Image.id.not_in([face.img_id for face in officer_faces])
+        ).first()
+
+        crop_mock = MagicMock(return_value=image)
+        upload_mock = MagicMock(return_value=image)
```
Comment (AetherUnbound) [inline OpenOversight/tests/routes/test_officer_and_department.py:1847]: Should we make this a test util function? I recognize that there's already a ton of duplication in the tests so feel free to ignore 😂

File: OpenOversight/tests/test_utils.py [hunk 5]
```diff
@@ -305,3 +309,29 @@ def test_crop_image_calls_upload_image_to_s3_and_store_in_db_with_user_id(
             assert (
                 current_user.get_id() in upload_image_to_s3_and_store_in_db.call_args[0]
             )
+
+
+@pytest.mark.parametrize(
+    "units, has_officers_with_unit, has_officers_with_no_unit",
+    [
+        (["Not Sure"], False, True),
+        (["Donut Devourers"], True, False),
```
Comment (AetherUnbound) [inline OpenOversight/tests/test_utils.py:318]: 💯 🍩 🚮

File: OpenOversight/tests/test_utils.py [hunk 6]
```diff
@@ -305,3 +309,29 @@ def test_crop_image_calls_upload_image_to_s3_and_store_in_db_with_user_id(
...
+
+    officers = filter_by_form(form_data, Officer.query, department_id).all()
+
+    for officer in officers:
+        found = False
+        if has_officers_with_unit:
+            found = found or any([a.unit_id == unit_id for a in officer.assignments])
+        if has_officers_with_no_unit:
+            found = found or any([a.unit_id is None for a in officer.assignments])
+        assert found
```
Comment (AetherUnbound) [inline OpenOversight/tests/test_utils.py:337]: Hmm, these checks make me a little concerned - in the case where both are `True`, the second conditional is totally ignored I think. Is that intended? Should we maybe set `found = True` at the beginning, then do `found = found and ...` (or something similar) so that the conditions must be met?

**Review by AetherUnbound — COMMENTED** [review]
Ah, this one's a bit trickier - nice find!

File: OpenOversight/tests/test_utils.py (hunk 6 above)
Comment (sea-kelp) [reply OpenOversight/tests/test_utils.py:337]: When we select 2 units (ex: "Donut Devourers" and "Bureau of Organized Crime") we are looking for any officers who have at least one assignment with either the "Donut Devourerers" or "Bureau of Organized Crime" unit.

This test is checking that same case but with "Donut Devourers" and "Not Sure". It loops through the list of `officers` and asserts that each officer either `has_officers_with_unit` or `has_officers_with_no_unit`. So the second conditional would be called for any officers who weren't "Donut Devourers"

File: OpenOversight/tests/test_utils.py (hunk 5 above)
Comment (sea-kelp) [reply OpenOversight/tests/test_utils.py:318]: The test data is fantastic
https://github.com/OrcaCollective/OpenOversight/blob/ca674ef880b91f6df1738287b3b13e76e94f270f/OpenOversight/tests/conftest.py#L302-L308

File: OpenOversight/app/utils.py (hunk 2 above)
Comment (sea-kelp) [reply OpenOversight/app/utils.py:365]: It totally makes sense to think that sqlalchemy would automagically handle querying a list containing None/null but sql is hell so :shrug:

File: OpenOversight/app/main/views.py (hunk 1 above)
Comment (sea-kelp) [reply OpenOversight/app/main/views.py:676]: Not exactly! "Not Sure" used to be added to the unit options list further down in the code:
https://github.com/OrcaCollective/OpenOversight/blob/ca674ef880b91f6df1738287b3b13e76e94f270f/OpenOversight/app/main/views.py#L718

But that meant that `form_data["unit"]` wouldn't be set if "Not Sure" was selected since it wouldn't be in `unit_choices`. (`form_data["unit"]` is used later on to construct the search query and re-populate the input fields)
https://github.com/OrcaCollective/OpenOversight/blob/ca674ef880b91f6df1738287b3b13e76e94f270f/OpenOversight/app/main/views.py#L690-L693

And that's basically what caused all options to become deselected if "Not Sure" was selected (like in the "Before" screenshot)

File: OpenOversight/tests/test_utils.py (hunk 6 above)
Comment (AetherUnbound) [reply OpenOversight/tests/test_utils.py:337]: Ahh got it! Yea I guess we only need to check that a value is returned 👍🏼

**Review by AetherUnbound — APPROVED** [review]
This looks great!
